atomicfu/src/concurrentTest/kotlin/kotlinx/atomicfu/parking/TimedParkingTest.kt

```diff
@@ -0,0 +1,229 @@
+
+        val thread1 = Fut {
+            kthread1 = KThread.currentThread()
+            val t = measureTime {
+                Parker.parkNanos(Long.MAX_VALUE / 3)
+            }
+            assertTrue(t.inWholeNanoseconds > 900_000_000)
+            assertTrue(t.inWholeNanoseconds < 1100_000_000)
+        }
+
+        sleepMills(1000)
+        Parker.unpark(kthread1!!)
+
+        thread1.waitThrowing()
+    }
+
+    @Test
+    fun testNanosFirstDeadline400() = retry(3) {
+        var kthread1: KThread? = null
+
+        val thread1 = Fut {
+            kthread1 = KThread.currentThread()
+            val t = measureTime {
+                Parker.parkNanos(400_000_000)
+            }
+            assertTrue(t.inWholeNanoseconds > 300_000_000)
+            assertTrue(t.inWholeNanoseconds < 500_000_000)
+        }
+
+        sleepMills(600)
+        Parker.unpark(kthread1!!)
+
+        thread1.waitThrowing()
+    }
+
+    @Test
+    fun testNanosFirstDeadline700() = retry(3) {
+        var kthread1: KThread? = null
+
+        val thread1 = Fut {
+            kthread1 = KThread.currentThread()
+            val t = measureTime {
+                Parker.parkNanos(700_000_000)
+            }
+            assertTrue(t.inWholeNanoseconds > 600_000_000)
+            assertTrue(t.inWholeNanoseconds < 800_000_000)
+        }
+
+        sleepMills(900)
+        Parker.unpark(kthread1!!)
+
+        thread1.waitThrowing()
+    }
+
+    @Test
+    fun testNanosFirstDeadline1200() = retry(3) {
+        var kthread1: KThread? = null
+
+        val thread1 = Fut {
+            kthread1 = KThread.currentThread()
+            val t = measureTime {
+                Parker.parkNanos(1000_000_000)
+            }
+            assertTrue(t.inWholeNanoseconds > 900_000_000)
+            assertTrue(t.inWholeNanoseconds < 1100_000_000)
+        }
+
+        sleepMills(1200)
+        Parker.unpark(kthread1!!)
+
+        thread1.waitThrowing()
+    }
+
+    private fun retry(times: Int, block: () -> Unit): Unit {
+        var lastThrowable: Throwable? = null
+        repeat(times) {
+            try {
+                return block()
+            } catch (t: Throwable) {
+                lastThrowable = t
+            }
+        }
+        val err = IllegalStateException("Failed after $times retries")
+        err.addSuppressed(lastThrowable?: IllegalStateException("Retry failed but no error"))
+        throw err
+    }
+}
+
+
+internal class Fut(private val block: () -> Unit) {
+    private var thread: TestThread? = null
+    private val atomicError = atomic<Throwable?>(null)
+    val done = atomic(false)
+    init {
+        val th = testThread {
+            try { block() }
+            catch (t: Throwable) {
+                atomicError.value = t
+                throw t
+            }
+            finally { done.value = true }
+        }
+        thread = th
+    }
+    fun waitThrowing() {
+        thread!!.join()
+        throwIfError()
+    }
```

Comment:
Minor, simplification:
```suggestion
internal class Fut(block: () -> Unit) {
    private val done = atomic(false)
    private val atomicError = atomic<Throwable?>(null)
    private val thread: TestThread = testThread {
        try { block() }
        catch (t: Throwable) {
            atomicError.value = t
            throw t
        }
        finally { done.value = true }
    }

    fun waitThrowing() {
        thread.join()
        throwIfError()
    }

```